### PR TITLE
Check modifiers on record positional members

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/45008
Corresponding spec update: https://github.com/dotnet/csharplang/pull/3672

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 1]
```diff
@@ -986,7 +986,7 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                         body = (BoundBlock)body.WithHasErrors();
                     }
 
-                    if (body != null && methodSymbol.IsConstructor())
+                    if (body != null && methodSymbol.IsConstructor() && !(methodSymbol is SynthesizedRecordConstructor))
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs:989]: > && !(methodSymbol is SynthesizedRecordConstructor) [](start = 68, length = 51)

Could you please provide some details about this change? I would not expect us to suppress any analysis for primary constructor.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs [hunk 2]
```diff
@@ -72,6 +72,8 @@ protected sealed override void MethodChecks(DiagnosticBag diagnostics)
 
 #nullable enable
         protected abstract ParameterListSyntax GetParameterList();
+
+        protected virtual bool AllowRefOrOut => true;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:76]: >protected virtual bool AllowRefOrOut => true; [](start = 8, length = 45)

abstract please

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs [hunk 3]
```diff
@@ -36,13 +33,11 @@ internal RecordDeclarationSyntax GetSyntax()
...
-            get
-            {
-                return false;
-            }
-        }
+        protected override bool AllowRefOrOut => false;
+
+        internal override bool IsExpressionBodied => false;
+
+        public override bool IsImplicitlyDeclared => true;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs:40]: >public override bool IsImplicitlyDeclared => true; [](start = 8, length = 50)

Please revert this change. This method is not implicitly declared. We had discussion about this in contex of the PR that changed base type for this class.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs [hunk 4]
```diff
@@ -98,7 +100,7 @@ public sealed override bool IsVararg
             }
         }
 
-        public sealed override bool IsImplicitlyDeclared
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceConstructorSymbolBase.cs:101]: >public sealed override bool IsImplicitlyDeclared [](start = 8, length = 48)

It looks like this change is unnecessary, given the comment in SynthesizedRecordConstructor.cs.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 5]
```diff
@@ -15638,5 +15638,213 @@ public void SynthesizedRecordPointerProperty()
...
+record R(in int P1);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+
+            var actualMembers = comp.GetMember<NamedTypeSymbol>("R").GetMembers().ToTestDisplayStrings();
+            var expectedMembers = new[]
+            {
+                "R R.<>Clone()",
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:15673]: >"R R.<>Clone()", [](start = 16, length = 16)

This is going to break once the name change for the Clone method is merged. In order to avoid unnecessary noise and maintainability burden, consider doing targeted checks only for members that you are interested in.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 6]
```diff
@@ -15638,5 +15638,213 @@ public void SynthesizedRecordPointerProperty()
...
+record R(params int[] Array);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
+
+            var actualMembers = comp.GetMember<NamedTypeSymbol>("R").GetMembers().ToTestDisplayStrings();
+            var expectedMembers = new[]
+            {
+                "R R.<>Clone()",
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:15718]: >"R R.<>Clone()", [](start = 16, length = 16)

Same comment.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 7]
```diff
@@ -15638,5 +15638,213 @@ public void SynthesizedRecordPointerProperty()
...
+
+        [Fact, WorkItem(45008, "https://github.com/dotnet/roslyn/issues/45008")]
+        public void PositionalMemberModifiers_In()
+        {
+            var src = @"
+record R(in int P1);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:15668]: >comp.VerifyDiagnostics(); [](start = 12, length = 25)

I think we should run and observe expected behavior at run-time.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 8]
```diff
@@ -15638,5 +15638,213 @@ public void SynthesizedRecordPointerProperty()
...
+
+        [Fact, WorkItem(45008, "https://github.com/dotnet/roslyn/issues/45008")]
+        public void PositionalMemberModifiers_Params()
+        {
+            var src = @"
+record R(params int[] Array);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:15713]: >comp.VerifyDiagnostics(); [](start = 12, length = 25)

I think we should run and observe expected behavior at run-time.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 9]
```diff
@@ -15638,5 +15638,213 @@ public void SynthesizedRecordPointerProperty()
...
+
+        [Fact, WorkItem(45008, "https://github.com/dotnet/roslyn/issues/45008")]
+        public void PositionalMemberModifiers_RefOrOut()
+        {
+            var src = @"
+record R(ref int P1, out int P2);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:15650]: >VerifyDiagnostics [](start = 17, length = 17)

VerifyEmitDiagnostics?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 10]
```diff
@@ -15638,5 +15638,213 @@ public void SynthesizedRecordPointerProperty()
...
+
+        [Fact, WorkItem(45008, "https://github.com/dotnet/roslyn/issues/45008")]
+        public void PositionalMemberModifiers_This()
+        {
+            var src = @"
+record R(this int i);
+";
+
+            var comp = CreateCompilation(src);
+            comp.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:15698]: >VerifyDiagnostics [](start = 17, length = 17)

VerifyEmitDiagnostics?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs:40]: Also, I believe this change is way outside of the advertised goal of this PR. 

---
In reply to: [454451705](https://github.com/dotnet/roslyn/pull/45898#discussion_r454451705) [](ancestors = 454451705)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs (hunk 3 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs:40]: That seems wrong to me, but if it's already been discussed and resolved then it's okay. Where did the discussion happen? 
I didn't see it in [https://github.com/dotnet/roslyn/pull/44803](https://github.com/dotnet/roslyn/pull/44803)

---
In reply to: [454465241](https://github.com/dotnet/roslyn/pull/45898#discussion_r454465241) [](ancestors = 454465241,454451705)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs:40]: >That seems wrong to me, but if it's already been discussed and resolved then it's okay. Where did the discussion happen? 
I didn't see it in [https://github.com/dotnet/roslyn/pull/44803](https://github.com/dotnet/roslyn/pull/44803)

You are correct, it is quite possible I confused it with this thread [https://github.com/dotnet/roslyn/pull/44803#pullrequestreview-423731255](https://github.com/dotnet/roslyn/pull/44803#pullrequestreview-423731255). In any case, there is a syntax explicitly declaring the constructor - at least the parameter list. There is also other user code in it - base arguments. Likely there will be a body in the future as well. Even today we should fire analyzers actions for the constructor, due to presence of user code. Analyzer driver filters out implicitly declared symbols all over the code.

---
In reply to: [454547147](https://github.com/dotnet/roslyn/pull/45898#discussion_r454547147) [](ancestors = 454547147,454465241,454451705)

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: It's not essential (optimization). I'll remove

---
In reply to: [454449395](https://github.com/dotnet/roslyn/pull/45898#discussion_r454449395) [](ancestors = 454449395)

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs [hunk 11]
```diff
@@ -41,7 +41,7 @@ internal class FlowAnalysisPass
             if (method.ReturnsVoid || method.IsIterator || method.IsAsyncReturningTask(compilation))
             {
                 // we don't analyze synthesized void methods.
-                if ((method.IsImplicitlyDeclared && !method.IsScriptInitializer) ||
+                if ((method.IsImplicitlyDeclared && !method.IsScriptInitializer) || method is SynthesizedRecordConstructor ||
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs:44]: :memo: This is primarily to avoid cascading diagnostics (illustrated below), but also seems reasonable (why do extra analysis work on synthesized methods which we fully control?).

```
        [Fact, WorkItem(45008, "[https://github.com/dotnet/roslyn/issues/45008](https://github.com/dotnet/roslyn/issues/45008)")]
        public void PositionalMemberModifiers_RefOrOut()
        {
            var src = @"
record R(ref int P1, out int P2);
";

            var comp = CreateCompilation(src);
            comp.VerifyDiagnostics(
                // (2,9): error CS0177: The out parameter 'P2' must be assigned to before control leaves the current method
                // record R(ref int P1, out int P2);
                Diagnostic(ErrorCode.ERR_ParamUnassigned, "(ref int P1, out int P2)").WithArguments("P2").WithLocation(2, 9),
                // (2,10): error CS0631: ref and out are not valid in this context
                // record R(ref int P1, out int P2);
                Diagnostic(ErrorCode.ERR_IllegalRefParam, "ref").WithLocation(2, 10),
                // (2,22): error CS0631: ref and out are not valid in this context
                // record R(ref int P1, out int P2);
                Diagnostic(ErrorCode.ERR_IllegalRefParam, "out").WithLocation(2, 22)
                );
        }
```

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs:44]: > why do extra analysis work on synthesized methods which we fully control?

We are not fully in control, there is user code in this method. Please revert this change. If you want to suppress specific diagnostics, suppress that specific diagnostics instead. It will be more robust to avoid making suppressions like that across the board. I personally, don't even think the diagnostics is cascading or should be suppressed, it reflects exactly what is happening, the parameter could be assigned in any of the initializers, or in base arguments.  

---
In reply to: [454571864](https://github.com/dotnet/roslyn/pull/45898#discussion_r454571864) [](ancestors = 454571864)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs:44]: We might even come up with a scenario where the analysis produces useful diagnostics about code in initializers, or in base arguments.

---
In reply to: [455167934](https://github.com/dotnet/roslyn/pull/45898#discussion_r455167934) [](ancestors = 455167934,454571864)

File: src/Test/Utilities/Portable/Diagnostics/DiagnosticExtensions.cs [hunk 12]
```diff
@@ -104,6 +104,14 @@ public static TCompilation VerifyDiagnostics<TCompilation>(this TCompilation c,
             return c;
         }
 
+        public static TCompilation VerifyDiagnosticsAndEmitDiagnostics<TCompilation>(this TCompilation c, params DiagnosticDescription[] expected)
+            where TCompilation : Compilation
+        {
+            VerifyDiagnostics(c, expected);
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Diagnostics/DiagnosticExtensions.cs:110]: >VerifyDiagnostics(c, expected); [](start = 12, length = 31)

This feels redundant, but, if we are willing to "pay" for the time it takes to run tests, we can do that.

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs (hunk 11 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs:44]: >method is SynthesizedRecordConstructor [](start = 84, length = 38)

I think this change should be reverted, see my comment on the ":memo:" thread.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 5)

File: src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs (hunk 11 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/FlowAnalysisPass.cs:44]: Thanks. I understand now. We definitely want to analyze the base call.

File: src/Test/Utilities/Portable/Diagnostics/DiagnosticExtensions.cs (hunk 12 above)
Comment (jcouv) [reply src/Test/Utilities/Portable/Diagnostics/DiagnosticExtensions.cs:110]: We've been multiplying tests that check both `VerifyDiagnostics` and `VerifyEmitDiagnostics`, so this is just making it easier/shorter to write such tests.

Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Diagnostics/DiagnosticExtensions.cs:110]: > We've been multiplying tests that check both VerifyDiagnostics and VerifyEmitDiagnostics, so this is just making it easier/shorter to write such tests.

I am not sure if that is really needed. I mean checking both, I was always suggesting to use one instead of the other. But I am fine with that as long as we are willing to wait longer for tests to complete.

---
In reply to: [455188087](https://github.com/dotnet/roslyn/pull/45898#discussion_r455188087) [](ancestors = 455188087)

Comment (jcouv) [reply src/Test/Utilities/Portable/Diagnostics/DiagnosticExtensions.cs:110]: Okay. Switched to use `VerifyEmitDiagnostics` only.
